iree/compiler/Conversion/LinalgToLLVM/LinalgBufferizePass.cpp

```diff
@@ -0,0 +1,447 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//===- LinalgBufferizePass.cpp.cpp - Pass to bufferize Linalg on tensors --===//
+//
+// Pass to convert from Linalg ops on tensors to Linalg ops on buffers.
+// This just inserts AllocOp to address space 0 that can be later hoisted,
+// promoted and generally rewritten to the desired backend.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Conversion/CodegenUtils/FunctionUtils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/IREE/IR/IREEDialect.h"
+#include "iree/compiler/Dialect/IREE/IR/IREEOps.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+static Value maybeConvertToIndex(Location loc, Value val, OpBuilder &b) {
+  if (val.getType().isIndex()) return val;
+  return b.create<IndexCastOp>(loc, val, b.getIndexType());
+}
+
+static LogicalResult allocateBuffersForResults(
+    OpBuilder &b, Location loc, linalg::LinalgOp op,
+    SmallVectorImpl<Value> &resultBuffers, BlockAndValueMapping &bvm) {
+  // Lazily compute loopRanges.
+  SmallVector<Range, 4> loopRanges;
+
+  for (auto en : llvm::enumerate(op.getOperation()->getResultTypes())) {
+    size_t resultIndex = en.index();
+    Type resultType = en.value();
+
+    auto tensorType = resultType.dyn_cast<RankedTensorType>();
+    if (tensorType == nullptr) {
+      op.emitOpError()
+          << "tensor to buffer conversion expects ranked tensor results";
+      return failure();
+    }
+    auto tensorShape = tensorType.getShape();
+    auto memrefType = MemRefType::get(tensorShape, tensorType.getElementType());
+
+    // Allocate buffers for init tensors that are assumed to fold onto the first
+    // results.
+    // TODO: update this assumption because the reality is more complex
+    // under linalg on tensor based transformations.
+    bool foldedInitTensor = resultIndex < op.getNumInitTensors();
+    if (foldedInitTensor) {
+      Value initTensor = op.getInitTensor(resultIndex);
+      if (initTensor.use_empty()) continue;
+      Value initBuffer = bvm.lookup(initTensor);
+      SmallVector<Value, 4> dynOperands;
+      for (auto dim : llvm::enumerate(tensorShape)) {
+        if (dim.value() == TensorType::kDynamicSize) {
+          dynOperands.push_back(b.create<DimOp>(loc, initTensor, dim.index()));
+        }
+      }
+      auto alloc = b.create<AllocOp>(loc, memrefType, dynOperands);
```

Comment:
Are you sure Alloc works? Maybe you meant to use Alloca?

Reply:
yes, I use ```
    passManager.addPass(createCopyRemovalPass());
    passManager.addPass(createBufferHoistingPass());
    passManager.addPass(createBufferLoopHoistingPass());
    passManager.addPass(createPromoteBuffersToStackPass(1 << 10, 64, 10));```
in my e2e branch.

This should be target-independent with a rewrite of the AllocOp as was previously proposed by @silvasean .
Will update if needed when we get to GPU.

Reply:
Ok cool. For our use case copy removal pass and buffer hoisting pass etc. might be fine. My experience in the past with such approaches has been less than ideal. They should work in theory well, but they get tripped up many times. Without control flow though a lot of the gotchas go away.